File: Trees/2331. Evaluate Boolean Binary Tree/2331.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def evaluateTree(self, root: Optional[TreeNode]) -> bool:
        def dfs(node):
            if not node:
                return
            if node.left and node.right:
                left=dfs(node.left)
                right=dfs(node.right)
            
            if not node.left and not node.right:
                return False if node.val==0 else True
            
            if node.val==2:
                return left or right
            elif node.val==3:
                return left and right
        
        return dfs(root)

File: Trees/2331. Evaluate Boolean Binary Tree/2331.cpp
#pragma GCC optimize("O3","Oz","unroll-loops","inline-functions","lto","fast-math","march=native","omit-frame-pointer")
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool evaluateTree(TreeNode* root) {
        ios_base::sync_with_stdio(0);
        cin.tie(0);
        cout.tie(0);
        return dfs(root);
    }
private:
    bool dfs(TreeNode* node) {
        if(!node->left && !node->right){
            if(node->val==0) return false;
            if(node->val==1) return true;
        }
        bool left=dfs(node->left);
        bool right=dfs(node->right);
        if(node->val==2) return left || right;
        return left && right;
    }
};